### PR TITLE
Adds compact

### PR DESCRIPTION
**Pull request by dgilperez**
Compact is heavily inspired by Ruby / Rails counterpart, that supports both [arrays](https://apidock.com/ruby/v2_6_3/Array/compact) and [hashes](https://apidock.com/rails/Hash/compact). 

I find it one most useful when dealing with real world data arrays and hashes. I made it ignore other types. 

Let me know what do you think or if you need any changes on the code.

### CODE REVIEW

File: src/compact/compact.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { forEach } from "../for-each/for-each"
+import { is } from "../is/is"
+import { map } from "../map/map"
+import { filter } from "../filter/filter"
+import { curry } from "../curry/curry"
+
+const _compactObject = source => {
```
Comment (andreidmt) [inline src/compact/compact.js:7]: Would skip here adding one more function call, `forEach`, and use a simple `for`:

```js
const _compactObject = source => {
  const result = { ...source }
  const keys = Object.keys(result)

  for (const key of keys) {
    if (!is(result[key])) {
      delete result[key]
    }
  }

  return result
}
```

File: src/compact/compact.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    }
+  }, Object.keys(result))
+
+  return result
+}
+
+const _compactArray = source => filter(is, source)
+
+const _compact = source => {
+  const type = Array.isArray(source) ? "array" : typeof source
```
Comment (andreidmt) [inline src/compact/compact.js]: There is a problem here with `typeof source`, when `null` is passed, the type is "object".

We already have a function, [`type`](https://github.com/asd-xiv/m/blob/master/src/type/type.js), that returns a more detailed and reliable description.

```js
import { type } from "../type/type"

const _compact = source => {
  const sourceType = type(source)

  switch (sourceType) {
    case "Array":
      return _compactArray(source)

    case "Object":
      return _compactObject(source)

    default:
      return source
  }
}
```

File: src/compact/compact.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    case "object":
+      return _compactObject(source)
+
+    default:
+      return source
+  }
+}
+
+const _compactMany = source => map(item => _compact(item), source)
```
Comment (andreidmt) [inline src/compact/compact.js]: This can be rewritten simply as:

```js
const _compactMany = map(_compact)
```

File: src/compact/compact.test.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
+/* eslint-disable unicorn/no-null */
+import test from "tape"
+
+import { compact, compactMany } from "./compact"
+
+const lambda = () => null
+
+test("compact", t => {
```
Comment (andreidmt) [inline src/compact/compact.test.js:8]: One test that is missing, testing for `null`, would of exposed the `typeof` error.

Either add it to the `compactMany([{}, 2, null, undefined])` that will proxy to the core `_compact` or add an explicit `compact(null)`.

I should write at some point a DoD I guess, to always test `null` and `undefined`, `*Many` and `*With` patterns, etc

File: src/compact/compact.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+export const compact = curry(_compact)
```
Comment (andreidmt) [inline src/compact/compact.js]: You don't need `curry` since `compact` and `compactMany` signatures have only one parameter.

Also, a comment block would help a lot, for both `compact` and `compactMany`, here's a starter:

```js
/**
 * Remove all "nothing" values from arrays or objects.
 *
 * @param {Array|Object} source
 *
 * @returns {Array|Object}
 *
 * @name compact
 * @tag Array,Object
 * @signature (source: Array|Object): Array|Object
 *
 * @see {@link compactMany}
 * @see {@link is}
 *
 * @example
 * compact([1, null, {}])
 * // => [1, {}]
 *
 * compact({
 *   a: "Lorem Ipsum",
 *   b: null,
 *   c: undefined,
 *   d: false,
 *   f: lambda,
 *  }),
 * // => {
 * //  a: "Lorem Ipsum",
 * //  d: false,
 * //  f: lamda
 * // }
 */
```

**Review by andreidmt — CHANGES_REQUESTED** [review]
Hey, yes, an `is` minigun for objects and arrays is a welcome addition.

Thank you for adding this, please see my comments.

**Comment by dgilperez** [comment]
Thanks for the quick review, all changes added.

**Comment by andreidmt** [comment]
Looks good, there's just the `curry` issue mentioned [here](https://github.com/asd-xiv/m/pull/3#discussion_r602696278) and it's ready.

**Comment by dgilperez** [comment]
Done!

**Comment by dgilperez** [comment]
Thanks! 🍻

>
